system-test/src/systemTest/java/org/apache/mesos/elasticsearch/systemtest/ReconciliationSystemTest.java

```diff
@@ -0,0 +1,204 @@
+        } else {
+            LOGGER.debug("Local docker environment");
+            innerDockerHost = CLUSTER.getMesosContainer().getIpAddress() + ":" + DOCKER_PORT;
+        }
+
+        DockerClientConfig.DockerClientConfigBuilder dockerConfigBuilder = DockerClientConfig.createDefaultConfigBuilder().withUri("http://" + innerDockerHost);
+
+        innerDockerClient = DockerClientBuilder.getInstance(dockerConfigBuilder.build()).build();
+
+        LOGGER.debug("Injecting executor");
+        CLUSTER.injectImage("mesos/elasticsearch-executor");
+        await().atMost(TIMEOUT, TimeUnit.SECONDS).until(() -> CONFIG.dockerClient.listContainersCmd().exec().size() > 0); // Wait until mesos-local has started.
+        List<Container> containers = CONFIG.dockerClient.listContainersCmd().exec();
+
+        // Find the mesos-local container so we can do docker in docker commands.
+        mesosClusterId = "";
+        for (Container container : containers) {
+            if (container.getImage().contains(MESOS_LOCAL_IMAGE_NAME)) {
+                mesosClusterId = container.getId();
+                break;
+            }
+        }
+        LOGGER.debug("Mini-mesos container ID: " + mesosClusterId);
+    }
+
+    @After
+    public void after() {
+        CONTAINER_MANGER.stopAll();
+    }
+
+    @Test
+    public void ifSchedulerLostShouldReconcileExecutors() throws IOException {
+        ElasticsearchSchedulerContainer scheduler = startSchedulerContainer();
+        assertCorrectNumberOfExecutors();
+
+        // Stop and restart container
+        CONTAINER_MANGER.stopContainer(scheduler);
+        startSchedulerContainer();
+        assertCorrectNumberOfExecutors();
+    }
+
+    @Test
+    public void ifExecutorIsLostShouldStartAnother() throws IOException {
+        startSchedulerContainer();
+        assertCorrectNumberOfExecutors();
+
+        killOneExecutor();
+
+        // Should restart an executor, so there should still be three
+        assertCorrectNumberOfExecutors();
+
+    }
+
+    @Test
+    public void ifExecutorIsLostWhileSchedulerIsDead() throws IOException {
+        ElasticsearchSchedulerContainer scheduler = startSchedulerContainer();
+        assertCorrectNumberOfExecutors();
+
+        // Kill scheduler
+        CONTAINER_MANGER.stopContainer(scheduler);
+
+        // Kill executor
+        killOneExecutor();
+
+        startSchedulerContainer();
+        // Should restart an executor, so there should still be three
+        assertCorrectNumberOfExecutors();
+    }
+
+    private void assertCorrectNumberOfExecutors() throws IOException {
+        await().atMost(TIMEOUT, TimeUnit.SECONDS).until(() -> clusterPs().size() == CLUSTER_SIZE);
+        List<Container> result = clusterPs();
+        LOGGER.debug("Mini-mesos PS command = " + Arrays.toString(result.stream().map(Container::getId).collect(Collectors.toList()).toArray()));
```

Comment:
Quite a lot of going back and forth between string, arrays and lists. There's a collector for joining Strings

```
List<Integer> numbers = Arrays.asList(1, 2, 3, 4);
 String commaSeparatedNumbers = numbers.stream()
     .map(i -> i.toString())
     .collect(Collectors.joining(", "));
```